src/builder.rs

```diff
@@ -0,0 +1,884 @@
+//! Helpers for building virtual DOM nodes.
+
+use crate::{Attribute, Listener, Node, RootRender, VdomWeak};
+use bumpalo::Bump;
+
+/// A virtual DOM element builder.
+///
+/// Typically constructed with element-specific constructors, eg the `div`
+/// function for building `<div>` elements or the `button` function for building
+/// `<button>` elements.
+#[derive(Clone, Debug)]
+pub struct ElementBuilder<'a, Listeners, Attributes, Children>
+where
+    Listeners: 'a + AsRef<[Listener<'a>]>,
+    Attributes: 'a + AsRef<[Attribute<'a>]>,
+    Children: 'a + AsRef<[Node<'a>]>,
+{
+    bump: &'a Bump,
+    tag_name: &'a str,
+    listeners: Listeners,
+    attributes: Attributes,
+    children: Children,
+}
+
+impl<'a, Listeners, Attributes, Children> ElementBuilder<'a, Listeners, Attributes, Children>
+where
+    Listeners: 'a + AsRef<[Listener<'a>]>,
+    Attributes: 'a + AsRef<[Attribute<'a>]>,
+    Children: 'a + AsRef<[Node<'a>]>,
+{
+    /// Set the listeners for this element.
+    ///
+    /// You can use this method to customize the backing storage for listeners,
+    /// for example to use a fixed-size array instead of the default
+    /// dynamically-sized `bumpalo::collections::Vec`.
+    ///
+    /// Any listeners already added to the builder will be overridden.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// use dodrio::{builder::*, bumpalo::Bump};
+    ///
+    /// let b = Bump::new();
+    ///
+    /// // Create a `<div>` with a fixed-size array of two listeners.
+    /// let my_div = div(&b)
+    ///     .listeners([
+    ///         on(&b, "click", |root, vdom, event| {
+    ///             // ...
+    ///         }),
+    ///         on(&b, "dblclick", |root, vdom, event| {
+    ///             // ...
+    ///         }),
+    ///     ])
+    ///     .finish();
+    /// ```
+    #[inline]
+    pub fn listeners<L>(self, listeners: L) -> ElementBuilder<'a, L, Attributes, Children>
```

Comment:
One thing I notice which might be nice is to avoid the usage of fixed-size arrays wherever possible to reduce some indentation and otherwise help provide conveniences. For example to attach two listeners you'd call `listener` (or `on`) twice. 

I suspect though that you'll want to avoid using `Vec` wherever possible because reallocations cause memory to get wasted in a bump allocator scenario. I wonder though if instead of a `Vec`-like data structure internally everything could be linked lists?

Reply:
I wonder if a linked-list style approach could also remove all the type parameters?